### ISSUE TITLE
Theme development for gatsby-manor

### ISSUE THREAD

**Issue by ajayns** [gatsbymanor/gatsby-theme-kit#10]
I've checked out the docs for theme development, and they seem to be pretty vague on explanation, focusing more on how to quickly get started. 

I've discussed what I'd like to achieve here: https://github.com/gatsbyjs/gatsby/issues/4236#issuecomment-370181227 

So I'd like to know how it actually works. Are gatsby-manor themes starters you can switch from using the jay-cli? So if I want post forms and so on, is it possible with a gatsby-manor theme? Any examples of implementation would also help. 

Thanks!

**Comment by nodox** [comment]
TL;DR Gatsby Manor themes are not starters but they are swappable. Yes, you can post to servers from forms but you'll need to configure your components for this. `jay` implements theming by swapping files in the `src` directory with all the files supplemented by the theme. 

Thank you for opening an issue! Happy to talk you through how `jay` works and Gatsby Manor. Tell me more about what you want to achieve in greater details so I can better assist you. How did you get started? Where you are at? What do you want the end product to look like?

What did you think of the theme development [tutorial](https://www.gatsbymanor.com/docs/tutorials/theme-development)? What did you feel was vague on the website? 

How would you feel if we added some docs on the architecture of how we solve theming in Gatsby?

Sorry for all the questions. Looking to fully understand where you are coming from so I can give oyu information that will be the most useful. 

Cheers!

**Comment by ajayns** [comment]
@nodox I understand that they're swappable, as I've checked out the CLI and it works smooth.
This is in the case of using the theme-kit and switching between themes initially to see how they look (basically like checking out different starters). But in the case where I have an existing Gatsby project which uses a few plugins and already has content, it's not as easy. It seems to throw errors and push whatever content/stuff I've worked on into a themes/starter folder and add the theme starter.

So by saying swappable do you mean you can swap them initially while there is no content to check out how each works? If so, this should be specifically mentioned, so that it doesn't throw the user's existing code somewhere. Also as @KyleAMathews mentioned this swapping between themes in an existing project is not possible unless there's a predefined schema. 

And yes, adding docs on the arch of how you plan to solve the theming would really help both theme developers get a good idea of what they're working on as well users understand how it all goes together.

And no problem, thank for you such a good response from your side!

**Comment by nodox** [comment]
>  But in the case where I have an existing Gatsby project which uses a few plugins and already has content, it's not as easy. It seems to throw errors and push whatever content/stuff I've worked on into a themes/starter folder and add the theme starter.

To be honest, `jay` was not built to work with existing project. It was a feature I would come back to when the moment arose. Now would be the perfect opportunity to collaborate with you to figure out the use cases that would needed to make jay compatible with your project.

> So by saying swappable do you mean you can swap them initially while there is no content to check out how each works? If so, this should be specifically mentioned, so that it doesn't throw the user's existing code somewhere.

Again, you would be the first person to help make the tool more compatible with an existing codebases! 

> Also as @KyleAMathews mentioned this swapping between themes in an existing project is not possible unless there's a predefined schema.

Swapping is possible even with the current state of gatsby, the next version of `jay` will address this. Essentially you just need to package the themes with a little more information to make swapping less messy.

> And yes, adding docs on the arch of how you plan to solve the theming would really help both theme developers get a good idea of what they're working on as well users understand how it all goes together.

What types of specifics would you want addressed? The more questions you have the more through we can be in the explanation and give you more valuable information. 😄

**Comment by Jaikant** [comment]
I was trying to change a theme from photon to massively. Not sure, what I did wrong, I get the error 

```
Move Error:  { Error: EEXIST: file already exists, link 'themes/massively' -> 'src'
  errno: -17,
  code: 'EEXIST',
  syscall: 'link',
  path: 'themes/massively',
  dest: 'src' }
```

Was it because I did not do an unmount? Below is the sequence of steps I followed.
Btw, I just got my first site with gatsbymanor  running at https://mountainadvice.com
I am now attempting to change the theme on this for another site 

```
Jaikants-MacBook-Pro:HamptaTreks jai$ jay list
massively
Jaikants-MacBook-Pro:HamptaTreks jai$ ls src/
MountainAdvice.png	__e2e__			fonts			pages			utils
Photon.jsx		__tests__		images			styles
README.md		components		layouts			templates
Jaikants-MacBook-Pro:HamptaTreks jai$ ls -l src/
total 120
-rw-r--r--   1 jai  staff  53138 Mar  7 15:47 MountainAdvice.png
-rw-r--r--   1 jai  staff    198 Mar  7 15:47 Photon.jsx
-rw-r--r--   1 jai  staff    157 Mar  7 15:47 README.md
drwxr-xr-x   3 jai  staff    102 Mar  7 15:47 __e2e__
drwxr-xr-x   4 jai  staff    136 Mar  7 15:47 __tests__
drwxr-xr-x  15 jai  staff    510 Mar  7 15:47 components
drwxr-xr-x   8 jai  staff    272 Mar  7 15:47 fonts
drwxr-xr-x  30 jai  staff   1020 Mar  7 15:47 images
drwxr-xr-x   3 jai  staff    102 Mar  7 15:47 layouts
drwxr-xr-x   9 jai  staff    306 Mar  7 15:47 pages
drwxr-xr-x   8 jai  staff    272 Mar  7 15:47 styles
drwxr-xr-x   4 jai  staff    136 Mar  7 15:47 templates
drwxr-xr-x   5 jai  staff    170 Mar  7 15:47 utils
Jaikants-MacBook-Pro:HamptaTreks jai$ ls -l themes/
total 0
drwxr-xr-x  15 jai  staff  510 Mar  7 16:01 massively
drwxr-xr-x   8 jai  staff  272 Mar  7 15:47 starter
Jaikants-MacBook-Pro:HamptaTreks jai$ jay mount massively
Move Error:  { Error: EEXIST: file already exists, link 'themes/massively' -> 'src'
  errno: -17,
  code: 'EEXIST',
  syscall: 'link',
  path: 'themes/massively',
  dest: 'src' }
```

**Comment by Jaikant** [comment]
I did an `unmount` followed by a `mount` and the command did not error. But it seems, my template components in src/templates are no longer there.

```
Jaikants-MacBook-Pro:HamptaTreks jai$ jay unmount photon
Jaikants-MacBook-Pro:HamptaTreks jai$ jay mount massively
Jaikants-MacBook-Pro:HamptaTreks jai$ gatsby develop
success delete html and css files from previous builds — 0.016 s
success open and validate gatsby-config.js — 0.006 s
success copy gatsby files — 0.017 s
success onPreBootstrap — 0.727 s
success source and transform nodes — 0.547 s
success building schema — 0.307 s
success createLayouts — 0.008 s
success createPages — 0.027 s
⠁ 
Your site's "gatsby-node.js" created a page with a component that doesn't exist

{ path: '/expedition',
  component: '/Users/jai/work/guides/consulting/HamptaTreks/src/templates/trek-index.js',
  context: { data: { allMarkdownRemark: [Object] } } }

See the documentation for createPage https://www.gatsbyjs.org/docs/bound-action-creators/#createPage
```

**Comment by Jaikant** [comment]
So it seems, Jay put my old `src` into `themes/photon/`
I copied the template files back into `src/templates`
I then copied a few of my independent components, the template depends on into `src/components`

Weirdly after this, I am experienced the graphQL query on the theme failing. Not sure if I should  do it this way or just start a new project and copy the relevant files into it.

For now, I will start a new project, so I can isolate the cause of the graphql query failure.

**Comment by Jaikant** [comment]
The graphQL queries in the massively theme, had the `date` and `path` in the frontmatter. Which I did not have part of my markdown. Do we really need the `date` and `path` as part of the template? I would vote to remove it.

**Comment by Jaikant** [comment]
Ok, so I managed to change the themes. I had to swap in my `template`, `components` and `utils`.

**Comment by ajayns** [comment]
@nodox 

Yes I understand jay is still in development and yes, we can work on improving it and maybe even directly integrate with Gatsby for theming but then, for now, since that's not the case, I'd say it should be mentioned in the docs about currently how it is or works. 
Cause a user who's trying jay for the first time on his existing project (which would be the most popular case), won't be please with the cli throwing the code into a folder and replacing with the new theme's starter. I was just concerned about this.

**Comment by nodox** [comment]
> Btw, I just got my first site with gatsbymanor running at https://mountainadvice.com

@Jaikant thats awesome! I will happily share this on our Twitter page so everyone can see. What is your Twitter handle?

Seems like you got it working. Great! 

Although to @ajayns point the experience leaves a lot of room for improvement for Devs on existing projects. The next release will address this by add a type of boostrapping/capabilities check to ensure you have all the necessary requirements to integrate `jay` to existing `Gatsby` project. Right now as you correctly pointed its more optimized for brand new site generations.

Also the themes will come in a more packaged version such that @Jaikant doesn't run into that same problem where some themes have the certain frontmatter entries but others don't. 

I will add these to the ever growing list of things to do. 😅

**Comment by nodox** [comment]
As of `jay ^0.3.3` this is no longer an issue. 

@Jaikant your theme is now running on `v0.1.0` and should be upgraded to `v0.2.0` The best way to upgrade would be to start a new project and observe the new structure of the theme then copy over the files into the proper location. If you complete this, a migration guide, (or CLI command) would be awesome!

Thank you @ajayns as well.

**Comment by Jaikant** [comment]
@nodox it was great bootstrapping with the theme and some initial components. 
If the themes could be using Kyle Mathews typography and/or using css-in-js like emotion or glamor it would be awesome.
